Graph type update for Cypher 5 Constraints (#1395)
neo4j/docs-cypher: modules/ROOT/pages/constraints/create-constraints.adoc

:description: Information about creating, listing, and dropping Neo4j's constraints.
:page-aliases: constraints/examples.adoc

> include::https://raw.githubusercontent.com/neo4j-graphacademy/courses/main/asciidoc/courses/cypher-indexes-constraints/ad.adoc[]

Reviewer: we don't want to keep this include?

Reviewer 2: added a suggestion to add this line again (below)

> :description: Information about creating Neo4j's constraints.

Reviewer 2: ```suggestion
:description: Information about creating Neo4j's constraints.
include::https://raw.githubusercontent.com/neo4j-graphacademy/courses/main/asciidoc/courses/cypher-indexes-constraints/ad.adoc[]
```

======

.Create a constraint requiring `Book` nodes to have unique `isbn` properties
// tag::schema_constraints_property_uniqueness[]
[source, cypher]
----
CREATE CONSTRAINT book_isbn
FOR (book:Book) REQUIRE book.isbn IS UNIQUE
----
// end::schema_constraints_property_uniqueness[]

.Result
[source, queryresult]
----
Added 1 constraint.
----

> 

Reviewer 2: ```suggestion

[NOTE]
The detailed statistics view currently says `Unique constraints added: 1`.
It is updated to say `Node property uniqueness constraints added: 1` in Cypher 25.
```

[NOTE]
The detailed statistics view currently says `Unique constraints added: 1`.

> It is updated to say `Node property uniqueness constraints added: 1` in Neo4j 2025.01.

Reviewer: Is it odd to have it say `currently` since I can be on a release after 2025.01 and still use Cypher 5?
> The detailed statistics view **currently** says `Unique constraints added: 1`.
It is updated to say `Node property uniqueness constraints added: 1` in **Neo4j 2025.01**.

(this of course follows for the other places as well)

Reviewer 2: hm... does this capture the cases where the wording is updated then?
is it only the neo4j version? or a combination of the cypher and neo4j versions?
if it's only the neo4j version, then we could be clearer with the version ranges

Reviewer: I think it's only based on Neo4j versions

Reviewer 2: removed 'currently'

[NOTE]
The detailed statistics view currently says `Unique constraints added: 1`.

> It is updated to say `Node property uniqueness constraints added: 1` in Cypher 25.

Reviewer: Some of these that now say Cypher 25 was updated in general in 2025.01 (not really related to Cypher version, see this comment: https://github.com/neo4j/docs-cypher/pull/1395#discussion_r2394049543). Might be good if the notes instead compare Neo4j version 5.x and that instead of Cypher 5 and Cypher 25? 

Something like:
```
The detailed statistics view in the Neo4j 5 versions say `Unique constraints added: 1`.
It is updated to say `Node property uniqueness constraints added: 1` in Neo4j 2025.01.
```

(this applies to more than just this note, it goes for several of them, but not all as some were cypher version related as well)

Added 1 constraint.
----

> 

Reviewer 2: ```suggestion

[NOTE]
The detailed statistics view currently says `Relationship uniqueness constraints added: 1`.
It is updated to say `Relationship property uniqueness constraints added: 1` in Cypher 25.
```

======

> [role=label--enterprise-edition]

Reviewer: Let's keep it in Cypher 5
```suggestion
[role=label--enterprise-edition label--new-5.9]
```

Reviewer: I was under the general impression we still had these for Cypher 5 to cover the 5.x releases, but has that changed so we don't want to keep it since Cypher 5 was frozen 2025.06? 
(I'm adding this part to all comments that are relevant for the 5.x vs calVar updates, you don't need to answer all threads with the same info, was more to be able to separate them from the others that are cypher 5 vs cypher 25 changes)

======

> [[create-property-type-constraint-union-type]]

Reviewer: ```suggestion
[role=label--new-5.11]
[[create-property-type-constraint-union-type]]
```

Reviewer: I was under the general impression we still had these for Cypher 5 to cover the 5.x releases, but has that changed so we don't want to keep it since Cypher 5 was frozen 2025.06? 
(I'm adding this part to all comments that are relevant for the 5.x vs calVar updates, you don't need to answer all threads with the same info, was more to be able to separate them from the others that are cypher 5 vs cypher 25 changes)

For example, it is possible to have a property uniqueness and a property existence constraint on the same label/relationship type and property combination, though this would be the equivalent of having a node or relationship key constraint.
A more useful example would be to combine a property type and a property existence constraint to ensure that the property exists and has the given type.

[[create-a-constraint-if-not-exist]]
=== Handling existing constraints when creating a constraint

To avoid failing on existing constraints, `IF NOT EXISTS` can be added to the `CREATE` command.
This will ensure that no error is thrown and that no constraint is created if any other constraint with the given name, or another constraint on the same constraint type and schema, or both, already exists.
For property type constraints the property type also needs to be the same.

> Instead, an informational notification is returned showing the existing constraint which blocks the creation.

Reviewer: ```suggestion
As of Neo4j 5.17, an informational notification is instead returned showing the existing constraint which blocks the creation.
```

Reviewer: I was under the general impression we still had these for Cypher 5 to cover the 5.x releases, but has that changed so we don't want to keep it since Cypher 5 was frozen 2025.06? 
(I'm adding this part to all comments that are relevant for the 5.x vs calVar updates, you don't need to answer all threads with the same info, was more to be able to separate them from the others that are cypher 5 vs cypher 25 changes)

.Error message
[source, error]
----

> An equivalent constraint already exists, 'Constraint( id=5, name='sequels', type='RELATIONSHIP PROPERTY UNIQUENESS', schema=()-[:SEQUEL_OF {order}]-(), ownedIndex=4 )'.

Reviewer: ```suggestion
An equivalent constraint already exists, 'Constraint( id=5, name='sequels', type='RELATIONSHIP UNIQUENESS', schema=()-[:SEQUEL_OF {order}]-(), ownedIndex=4 )'.
```

Reviewer: I was under the general impression we still had these for Cypher 5 to cover the 5.x releases, but has that changed so we don't want to keep it since Cypher 5 was frozen 2025.06? 
(I'm adding this part to all comments that are relevant for the 5.x vs calVar updates, you don't need to answer all threads with the same info, was more to be able to separate them from the others that are cypher 5 vs cypher 25 changes)

----

> 

Reviewer 2: ```suggestion

[NOTE]
The constraint type is updated to say `RELATIONSHIP PROPERTY UNIQUENESS` in Cypher 25.
```

----

[NOTE]

> The constraint type is updated to say `RELATIONSHIP PROPERTY UNIQUENESS` in a future version of Neo4j.

Reviewer: Just realized that this was also updated in 2025.01 🤷 
```suggestion
The constraint type is updated to say `RELATIONSHIP PROPERTY UNIQUENESS` in Neo4j 2025.01.
```

.Error message
[source, error]
----

> Constraint already exists: Constraint( id=3, name='book_isbn', type='NODE PROPERTY UNIQUENESS', schema=(:Book {isbn}), ownedIndex=2 )

Reviewer: same here
```suggestion
Constraint already exists: Constraint( id=3, name='book_isbn', type='UNIQUENESS', schema=(:Book {isbn}), ownedIndex=2 )
```

and the note below

Reviewer: I was under the general impression we still had these for Cypher 5 to cover the 5.x releases, but has that changed so we don't want to keep it since Cypher 5 was frozen 2025.06? 
(I'm adding this part to all comments that are relevant for the 5.x vs calVar updates, you don't need to answer all threads with the same info, was more to be able to separate them from the others that are cypher 5 vs cypher 25 changes)

Constraint already exists: Constraint( id=3, name='book_isbn', type='UNIQUENESS', schema=(:Book {isbn}), ownedIndex=2 )
----

[NOTE]

> The constraint type is updated to say `NODE PROPERTY UNIQUENESS` in a future version of Neo4j.

Reviewer: same here
```suggestion
The constraint type is updated to say `NODE PROPERTY UNIQUENESS` in Neo4j 2025.01.
```

Constraint already exists: Constraint( id=3, name='book_isbn', type='UNIQUENESS', schema=(:Book {isbn}), ownedIndex=2 )
----

> ======

Reviewer: This note still looks removed :(
```suggestion
[NOTE]
The constraint type is updated to say `NODE PROPERTY UNIQUENESS` in Cypher 25.

======
```

.Error message
[source, error]
----

> Constraint already exists: Constraint( id=7, name='book_title_year', type='NODE PROPERTY UNIQUENESS', schema=(:Book {title, publicationYear}), ownedIndex=6 )

Reviewer: ```suggestion
Constraint already exists: Constraint( id=7, name='book_title_year', type='UNIQUENESS', schema=(:Book {title, publicationYear}), ownedIndex=6 )
```

Reviewer: I was under the general impression we still had these for Cypher 5 to cover the 5.x releases, but has that changed so we don't want to keep it since Cypher 5 was frozen 2025.06? 
(I'm adding this part to all comments that are relevant for the 5.x vs calVar updates, you don't need to answer all threads with the same info, was more to be able to separate them from the others that are cypher 5 vs cypher 25 changes)

.Error message
[source, error]
----

> Unable to create Constraint( name='book_title', type='NODE PROPERTY UNIQUENESS', schema=(:Book {title}) ):

Reviewer: ```suggestion
Unable to create Constraint( name='book_title', type='UNIQUENESS', schema=(:Book {title}) ):
```

Reviewer: I was under the general impression we still had these for Cypher 5 to cover the 5.x releases, but has that changed so we don't want to keep it since Cypher 5 was frozen 2025.06? 
(I'm adding this part to all comments that are relevant for the 5.x vs calVar updates, you don't need to answer all threads with the same info, was more to be able to separate them from the others that are cypher 5 vs cypher 25 changes)
